### ISSUE TITLE
Gradients for Loss function

### ISSUE THREAD

**Issue by DonghyunK** [wkentaro/pytorch-fcn#16]
Hi,

I have a question for loss function,
 
`loss = F.nll_loss(log_p, target, weight=weight, size_average=False)
    if size_average:
        loss /= mask.sum().data[0]`

Whether 'size_average' is True or False, scales of gradients for the loss function are different. 

This may be a problem because different scales of gradients may impact on how we set learning rate.

Is this the reason you use very low learning rate, 1e-10?

Thanks!!

**Comment by wkentaro** [comment]
Yeah, this kind of loss function's configurations are discussed in the original paper published in PAMI.

**Comment by DonghyunK** [comment]
Thank you for quick response!

Are you talking about the FCN paper published at PAMI? 

Could you please explain it more in detail?

I think it also matters with a batch size.

**Comment by wkentaro** [comment]
> Are you talking about the FCN paper published at PAMI?

Yeah. They compares performances with:

- loss accumulation about batch
- size_average=True
- size_average=False with batch_size=1

and they say the last one is best for VOC dataset.
But I think we should try all of them with other datasets.

**Comment by DonghyunK** [comment]
Thank you, but I am still very confusing.

Normally, a gradient of cross entropy with softmax is y^−y.

[http://stats.stackexchange.com/questions/149139/vectorization-of-cross-entropy-loss](url)

In the trainer.py, 

`loss = cross_entropy2d(score, target, size_average=False)`
`loss /= len(data)`

I think loss is divided by the size of batch.
Therefore, the gradient for cross entropy  will be  (y^−y)/len(data) which means that the gradient will be divided by its size of batch.

However, their FCN paper at PAMI, they said,

> 4.2 Image-to-image learning
The image-to-image learning setting includes high effective
batch size and correlated inputs. This optimization requires
some attention to properly tune FCNs.
We begin with the loss. We do not normalize the loss, so
that every pixel has the same weight regardless of the batch
and image dimensions. Thus we use a small learning rate
since the loss is summed spatially over all pixels.

Therefore, I think ` loss /= len(data)` is also should be removed 

Please, correct me if I am wrong and is this on purpose?

Thank you!
